From 18a2a8a29705412c675cb276625f239119131ddc Mon Sep 17 00:00:00 2001
From: Todd Whiteman <todd.whiteman@joyent.com>
Date: Sun, 10 Sep 2017 09:54:47 -0700
Subject: [PATCH] IMGAPI-644 the docker image cache must clear activated image
 entries Reviewed by: Trent Mick <trentm@gmail.com> Approved by: Trent Mick
 <trentm@gmail.com>

---
 CHANGES.md    | 4 ++++
 lib/docker.js | 7 +++++++
 package.json  | 2 +-
 3 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/CHANGES.md b/CHANGES.md
index e51c40d..277d000 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,5 +1,9 @@
 # IMGAPI changelog
 
+## 4.0.8
+
+- IMGAPI-644 the docker image cache must clear activated image entries
+
 ## 4.0.7
 
 - IMGAPI-642 missing file0 from imgapi docker image causing container provision
diff --git a/lib/docker.js b/lib/docker.js
index 6aa866b..564d32c 100644
--- a/lib/docker.js
+++ b/lib/docker.js
@@ -167,6 +167,10 @@ function dockerDownloadAndImportLayer(opts, callback) {
             log.debug({digest: digest, uuid: uuid},
                 'dockerDownloadAndImportImage: image layer already cached');
 
+            // IMGAPI-644 This cached image should not be activated.
+            assert.equal(newImage.activated, false,
+                'newImage.activated should be false');
+
             progressStatus('Download complete (cached)');
             callback(null, newImage);
             return;
@@ -715,6 +719,9 @@ function _dockerActivateImage(newImage, ctx, callback) {
                 return next(mErr);
             }
 
+            // The image file has been moved - remove it from the cache.
+            delete DOCKER_IMAGE_CACHE[newImage.uuid];
+
             newImage.addFile(app, newFileInfo.file, req.log, function (err2) {
                 if (err2) {
                     req.log.error(err2, 'error adding file info to Image');
diff --git a/package.json b/package.json
index f8bf4ae..21f37c1 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "imgapi",
   "description": "Image API to manage images for SDC 7",
-  "version": "4.0.7",
+  "version": "4.0.8",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
-- 
2.21.0

